File: go/euler052.go
// https://projecteuler.net/problem=52
//
// Permuted Multiples
//
// It can be seen that the number, 125874, and its double, 251748, contain exactly the same digits, but in a different order.
// Find the smallest positive integer, x, such that 2x, 3x, 4x, 5x, and 6x, contain the same digits.
//
// Answer: 142857

package main

import (
	"fmt"
	"sort"
	"strconv"
)

func hasSameDigits(a, b int) bool {
	sa := strconv.Itoa(a)
	sb := strconv.Itoa(b)
	if len(sa) != len(sb) {
		return false
	}
	da := []byte(sa)
	db := []byte(sb)
	sort.Slice(da, func(i, j int) bool { return da[i] < da[j] })
	sort.Slice(db, func(i, j int) bool { return db[i] < db[j] })
	return string(da) == string(db)
}

func main() {
	for x := 1; ; x++ {
		allSame := true
		for mul := 2; mul <= 6; mul++ {
			if !hasSameDigits(x, x*mul) {
				allSame = false
				break
			}
		}
		if allSame {
			fmt.Println(x)
			break
		}
	}
}

File: rust/src/euler052.rs
// https://projecteuler.net/problem=52
//
// It can be seen that the number, 125874, and its double, 251748, contain exactly the same digits, but in a different order.
// Find the smallest positive integer, x, such that 2x, 3x, 4x, 5x, and 6x, contain the same digits.
//
// Answer: 142857

pub fn permuted_multiples() -> u64 {
    for x in 1.. {
        let x_str = x.to_string();
        let mut digits: Vec<char> = x_str.chars().collect();
        digits.sort();
        let mut all_same = true;
        for mul in 2..=6 {
            let mx = x * mul;
            let mx_str = mx.to_string();
            if mx_str.len() != x_str.len() { all_same = false; break; }
            let mut mx_digits: Vec<char> = mx_str.chars().collect();
            mx_digits.sort();
            if mx_digits != digits { all_same = false; break; }
        }
        if all_same { return x; }
    }
    0
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn euler_052() {
        assert_eq!(permuted_multiples(), 142857);
    }
}